curp/src/sync_manager.rs

```diff
@@ -415,4 +444,225 @@ impl<C: Command + 'static> SyncManager<C> {
+                            break false;
+                        }
+                        ServerRole::Candidate => {}
+                    }
+                    if Instant::now() - *last_rpc_time.read() > Self::CANDIDATE_TIMEOUT {
+                        break true;
+                    }
+                },
+                ServerRole::Leader => false, // leader should not vote
+            };
+            if !start_vote {
+                continue;
+            }
+
+            // start election
+            #[allow(clippy::integer_arithmetic)] // TODO: handle possible overflow
+            let req = {
+                let mut state = state.write();
+                let new_term = state.term + 1;
+                state.update_to_term(new_term);
+                state.set_role(ServerRole::Candidate);
+                state.voted_for = Some(state.id.clone());
+                state.votes_received = 1;
+                VoteRequest::new(
+                    state.term,
+                    state.id.clone(),
+                    state.last_log_index(),
+                    state.last_log_term(),
+                )
+            };
+            // reset
+            *last_rpc_time.write() = Instant::now();
+            debug!("server {} starts election", req.candidate_id);
+
+            for connect in &connects {
+                let _ignore = tokio::spawn(Self::send_vote(
+                    Arc::clone(connect),
+                    Arc::clone(&state),
+                    req.clone(),
+                    Arc::clone(&calibrate_trigger),
+                ));
+            }
+        }
+    }
+
+    /// send vote request
+    async fn send_vote(
+        connect: Arc<Connect>,
+        state: Arc<RwLock<State<C>>>,
+        req: VoteRequest,
+        calibrate_trigger: Arc<Event>,
+    ) {
+        let resp = connect.vote(req, Self::RPC_TIMEOUT).await;
+        match resp {
+            Err(e) => error!("vote failed, {}", e),
+            Ok(resp) => {
+                let resp = resp.into_inner();
+
+                // calibrate term
+                let state = state.upgradable_read();
+                if resp.term > state.term {
+                    let mut state = RwLockUpgradableReadGuard::upgrade(state);
+                    state.update_to_term(resp.term);
+                    return;
+                }
+
+                // is still a candidate
+                if !matches!(state.role(), ServerRole::Candidate) {
+                    return;
+                }
+
+                #[allow(clippy::integer_arithmetic)]
+                if resp.vote_granted {
+                    debug!("vote is granted by server {}", connect.addr);
+                    let mut state = RwLockUpgradableReadGuard::upgrade(state);
+                    state.votes_received += 1;
+
+                    // the majority has granted the vote
+                    let min_granted = (state.others.len() + 1) / 2 + 1;
```

Comment:
if there are 3 servers, others = 2, min_grant = 2, the server can  becomes a leader only if votes_received = 3. Is it correct?

Reply:
Fixed